frontend/lib/algolia/server/findDescendantDevicesWithProducts.ts

```diff
@@ -0,0 +1,39 @@
+import {
+   ALGOLIA_API_KEY,
+   ALGOLIA_APP_ID,
+   ALGOLIA_PRODUCT_INDEX_NAME,
+} from '@config/env';
+import { CLIENT_OPTIONS, escapeFilterValue } from '@helpers/algolia-helpers';
+import { timeAsync } from '@ifixit/stats';
+import { cache } from '@lib/cache';
+import algoliasearch from 'algoliasearch';
+
+export async function findDescendantDevicesWithProducts(device: string) {
+   return cache(
+      `algolia.findDescendantDevicesWithProducts.${device}`,
+      () =>
+         timeAsync('algolia.findDescendantDevicesWithProducts', () =>
+            fetchDevices(device)
+         ),
+      60 * 10
+   );
+}
+
+async function fetchDevices(device: string) {
+   const client = algoliasearch(
+      ALGOLIA_APP_ID,
+      ALGOLIA_API_KEY,
+      CLIENT_OPTIONS
+   );
+   const index = client.initIndex(ALGOLIA_PRODUCT_INDEX_NAME);
+   const deviceSuffix = device
+      ? ` AND device:"${escapeFilterValue(device)}"`
```

Comment:
It seems like these two things aren't necessarily exhaustively overlapping:

* list of children devices of X
* list of devices from all products that are compatible with a descendant of X

It's possible that the second definitely contains all of the first, but I'm a little leery of that being 100% true and it would suck if some were left out.

Here's the kind of things that this query returns for `iphone`:
![image](https://user-images.githubusercontent.com/26855/212214075-f58a976e-535c-4ea6-b005-8e83f5974255.png)

I'm tempted to say we should do something more exact that involves our API but strapi product lists aren't tied directly to our device hierarchy...  so I'm not sure what to suggest. I'm nervous about depending on the list of facets being returned being completely exhaustive, especially when it's long. I feel like we should thoroughly review the differences here. Maybe just rendering them differently for now, or logging them or something.

Thoughts?

Reply:
One possible solution:

In Strapi the `deviceTitle` field only accepts one value.
![image](https://user-images.githubusercontent.com/52104630/212435178-1c110c6f-6c34-4baf-856d-37da84154937.png)
If this field (or another field like `relatedDeviceTitles`) accepted a list of values (for example, `[Appliance, Vehicle, etc]`), then we could better match on devices in algolia that don't have a product list entry in Strapi, but exist in the iFixit device hierarchy.

![image](https://user-images.githubusercontent.com/52104630/212436261-eeb371f2-a2e5-4a26-8f99-fdfde79cd3ce.png)
![image](https://user-images.githubusercontent.com/52104630/212436308-e165eaaf-18df-4b9a-a27a-57f32b636f9f.png)

Reply:
I'm not sure I understand your suggestion. Strapi product lists currently represent one since device. It seems like you're suggesting having them represent more than one? It seems like that would create far more problems than it would solve (multiple sets of ancestors, multiple sets of children, ...). Maybe I'm misunderstanding your suggestion.

Reply:
Danny and I talked, and came up with the solution implemented in https://github.com/iFixit/react-commerce/pull/1252/commits/f8b30b6e93545bbdf0768a8f3ec10fdb0d461685